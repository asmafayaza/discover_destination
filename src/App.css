html,
body {
    margin: 0px;
    height: 100%;
    overflow: hidden;
}

#root {
    height: inherit;
}

.list-unstyle {
    list-style: none;
}

.reduce-padding {
    padding-left: 10px;
    padding-right: 10px;
}
