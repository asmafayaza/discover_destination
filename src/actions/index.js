
export const searchFlight = (payload) => ({
  type: 'SEARCH_FLIGHT',
  payload
});

export const getLatestFlightResult = (searchString, flightList) => ({
  type: 'GET_LATEST_FLIGHT_RESULT',
  searchString,
  flightList
});
