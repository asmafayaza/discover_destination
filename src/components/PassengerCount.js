import React from "react";
import { ActionButtonWrapper } from "../styles/SearchStyle";

const PassengerCount = (props) => {
    const changePassengerCount = (e) => props.changePassengerCount(e);

    return (
        <ActionButtonWrapper>
            <div className="passenger-wrapper">
                <div className="passenger-title">
                    <h4 className="title">Passenger</h4>
                </div>
                <div className="passenger-button">
                    <button
                        id="reduceButton"
                        type="button"
                        className="button"
                        disabled={props.passengerCount <= props.minPassenger}
                        onClick={changePassengerCount}
                    >
                        -
                    </button>
                    <span className="passenger__count">
                        {props.passengerCount}{" "}
                    </span>
                    <button
                        id="increaseButton"
                        type="button"
                        className="button"
                        disabled={props.passengerCount >= props.maxPassenger}
                        onClick={changePassengerCount}
                    >
                        +
                    </button>
                </div>
            </div>
        </ActionButtonWrapper>
    );
};

export default PassengerCount;
