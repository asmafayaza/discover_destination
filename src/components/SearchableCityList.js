import React, { Component } from "react";
import EachCityItem from "./EachCityItem";
import PropTypes from "prop-types";

class SearchableCityList extends Component {
    constructor(props) {
        super(props);
        this.state = { selectedCity: "" };

        this.searchCityList = this.searchCityList.bind(this);
        this.selectCity = this.selectCity.bind(this);
        this.clearValue = this.clearValue.bind(this);
    }

    searchCityList(event) {
        this.setState({ selectedCity: event.target.value });
        this.props.searchCityList(event.target.value);
    }

    clearValue(event) {
        this.setState({ selectedCity: "" });
    }
    selectCity(city) {
        this.setState({ selectedCity: city.name });
        this.props.selectCity(city);
    }

    render() {
        const { selectedCity } = this.state;
        const { id, name, placeholder, showCityList, cityList } = this.props;

        return (
            <div className="container">
                <input
                    type="text"
                    name={name}
                    className="cityInput"
                    id={id}
                    value={selectedCity}
                    onChange={this.searchCityList}
                    placeholder={placeholder}
                    onBlur={this.clearValue}
                    autoComplete="true"
                    required
                />

                {showCityList && (
                    <ul className="list-unstyle city-list">
                        {cityList.length !== 0 &&
                            cityList.map((city) => (
                                <EachCityItem
                                    key={city.id}
                                    city={city}
                                    selectCity={this.selectCity}
                                />
                            ))}
                        {cityList.length === 0 && (
                            <li className="emptyMsg">No City Found</li>
                        )}
                    </ul>
                )}
            </div>
        );
    }
}

export default SearchableCityList;

SearchableCityList.propTypes = {
    cityList: PropTypes.array,
    searchCityList: PropTypes.func,
    selectCity: PropTypes.func,
    selectedCity: PropTypes.string,
    showCityList: PropTypes.bool,
};
