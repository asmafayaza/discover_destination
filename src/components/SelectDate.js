import React from "react";
import { DateField } from "../styles/SearchStyle";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import moment from "moment";

const SelectDate = (props) => {
    const changeDepartureDate = (date) => props.changeDepartureDate(date);
    const changeReturnDate = (date) => props.changeReturnDate(date);
    return (
        <DateField>
            <div className="wrapper">
                <h4 className="title">Departure date</h4>
                <DatePicker
                    className="inputDate"
                    required
                    minDate={moment()}
                    maxDate={moment().add(6, "months")}
                    selected={props.selectedDepartureDate}
                    onChange={changeDepartureDate}
                />
            </div>

            {props.roundTrip && (
                <div className="wrapper">
                    <h4 className="title">Return date</h4>
                    <DatePicker
                        className="inputDate"
                        required
                        minDate={props.selectedDepartureDate}
                        maxDate={moment().add(6, "months")}
                        selected={props.selectedReturnDate}
                        onChange={changeReturnDate}
                    />
                </div>
            )}
        </DateField>
    );
};

export default SelectDate;
