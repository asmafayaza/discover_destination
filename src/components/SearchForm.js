import React, { Component } from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import * as flightSearchAction from "../actions";
import CityList from "../CityList.json";
import SelectTrip from "./SelectTrip";
import SelectCity from "./SelectCity";
import SelectDate from "./SelectDate";
import PassengerCount from "./PassengerCount";
import RefinePrice from "./RefinePrice";
import { FlightSearch } from "../styles/SearchStyle";
import moment from "moment";

class SearchForm extends Component {
    constructor(props) {
        super(props);
        this.state = {
            originCityList: [],
            destinationCityList: [],
            showOriginCityList: false,
            showDestinationCityList: false,
            roundTrip: false,
            originCity: "",
            destinationCity: "",
            departureDate: moment(),
            returnDate: "",
            passengerCount: 1,
            price: {
                min: 1500,
                max: 10000,
            },
        };

        this.handleSearch = this.handleSearch.bind(this);
        this.selectOneWay = this.selectOneWay.bind(this);
        this.selectRoundTrip = this.selectRoundTrip.bind(this);
        this.changeDepartureDate = this.changeDepartureDate.bind(this);
        this.changeReturnDate = this.changeReturnDate.bind(this);
        this.changePassengerCount = this.changePassengerCount.bind(this);
        this.filterByPrice = this.filterByPrice.bind(this);
        this.searchOriginCity = this.searchOriginCity.bind(this);
        this.selectOriginCity = this.selectOriginCity.bind(this);
        this.searchDestinationCity = this.searchDestinationCity.bind(this);
        this.selectDestinationCity = this.selectDestinationCity.bind(this);
    }

    searchOriginCity(searchTerm) {
        const originCityListCopy = CityList.City.slice();

        const filterdList = originCityListCopy.filter((city) =>
            city.name.toLowerCase().includes(searchTerm.toLowerCase())
        );
        this.setState({
            showOriginCityList: true,
            showDestinationCityList: false,
            originCityList: filterdList,
            destinationCityList: [],
        });
    }

    selectOriginCity(city) {
        this.setState({
            originCityList: [],
            showOriginCityList: false,
            showDestinationCityList: false,
            originCity: city.code,
        });
    }

    searchDestinationCity(searchTerm) {
        const destinationCityListCopy = CityList.City.slice();
        const filterdList = destinationCityListCopy.filter((city) => {
            return city.name.toLowerCase().includes(searchTerm.toLowerCase());
        });

        this.setState({
            showOriginCityList: false,
            originCityList: [],
            showDestinationCityList: true,
            destinationCityList: filterdList,
        });
    }

    selectDestinationCity(city) {
        this.setState({
            destinationCityList: [],
            showOriginCityList: false,
            showDestinationCityList: false,
            destinationCity: city.code,
        });
    }

    selectOneWay(e) {
        this.setState({ roundTrip: false, returnDate: "" });
    }

    selectRoundTrip(e) {
        this.setState((prevState) => ({
            roundTrip: true,
            returnDate: moment(prevState.departureDate).add(1, "day"),
        }));
    }

    changeDepartureDate(date) {
        this.setState((prevState) => ({
            departureDate: date,
            returnDate: prevState.roundTrip ? moment(date).add(1, "day") : "",
        }));
    }

    changeReturnDate(date) {
        this.setState({
            returnDate: date,
        });
    }

    changePassengerCount(event) {
        if (event.target.id === "reduceButton") {
            this.setState((prevState) => ({
                passengerCount: prevState.passengerCount - 1,
            }));
        } else if (event.target.id === "increaseButton") {
            this.setState((prevState) => ({
                passengerCount: prevState.passengerCount + 1,
            }));
        }
    }

    filterByPrice(newPrice) {
        this.setState({ price: newPrice });
    }

    handleSearch(event) {
        const {
            roundTrip,
            originCity,
            destinationCity,
            departureDate,
            returnDate,
            passengerCount,
            price,
        } = this.state;

        if (originCity && destinationCity && departureDate) {
            const payload = {
                searchId: Date.now(),
                roundTrip: roundTrip,
                originCity: originCity,
                destinationCity: destinationCity,
                departureDate: departureDate,
                returnDate: returnDate,
                passengerCount: passengerCount,
                price: price,
            };
            this.props.actions.searchFlight(payload);
        }

        event.preventDefault();
    }

    render() {
        const {
            originCityList,
            destinationCityList,
            showOriginCityList,
            showDestinationCityList,
            roundTrip,
            originCity,
            destinationCity,
            departureDate,
            returnDate,
            passengerCount,
            price,
        } = this.state;

        return (
            <FlightSearch>
                <form
                    name="flightSearchForm"
                    className="searchForm"
                    onSubmit={this.handleSearch}
                >
                    <SelectTrip
                        selectOneWay={this.selectOneWay}
                        selectRoundTrip={this.selectRoundTrip}
                    />
                    <div className="content-container">
                        <div className="city-container">
                            <SelectCity
                                originCityList={originCityList}
                                destinationCityList={destinationCityList}
                                showOriginCityList={showOriginCityList}
                                showDestinationCityList={
                                    showDestinationCityList
                                }
                                selectedOriginCity={originCity}
                                selectedDestinationCity={destinationCity}
                                searchOriginCity={this.searchOriginCity}
                                searchDestinationCity={
                                    this.searchDestinationCity
                                }
                                selectOriginCity={this.selectOriginCity}
                                selectDestinationCity={
                                    this.selectDestinationCity
                                }
                            />
                        </div>

                        <div>
                            <div className="date-container">
                                <SelectDate
                                    selectedDepartureDate={departureDate}
                                    changeDepartureDate={
                                        this.changeDepartureDate
                                    }
                                    selectedReturnDate={returnDate}
                                    changeReturnDate={this.changeReturnDate}
                                    roundTrip={roundTrip}
                                />
                            </div>
                            <div className="filter-container">
                                <div className="passenger-input">
                                    <PassengerCount
                                        maxPassenger={6}
                                        minPassenger={1}
                                        changePassengerCount={
                                            this.changePassengerCount
                                        }
                                        passengerCount={passengerCount}
                                    />
                                </div>
                                <div className="price-input">
                                    <RefinePrice
                                        maxValue={15000}
                                        minValue={1000}
                                        price={price}
                                        filterByPrice={this.filterByPrice}
                                    />
                                </div>
                                <div className="btn-wrapper">
                                    <input
                                        type="submit"
                                        className="search-btn"
                                        value="Search"
                                        name="search_flight"
                                    />
                                </div>
                            </div>
                        </div>
                    </div>
                </form>
            </FlightSearch>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    actions: bindActionCreators(flightSearchAction, dispatch),
});

export default connect(null, mapDispatchToProps)(SearchForm);
