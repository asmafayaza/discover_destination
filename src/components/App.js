import React from "react";
import "../App.css";
import SearchForm from "./SearchForm";
import FlightList from "./FlightList";
import Header from "../components/Header/Header";
import Footer from "../components/Footer/footer";

function App() {
    return (
        <div className="App">
            <Header />
            <div className="Content">
                <SearchForm />
                <FlightList />
            </div>
            <Footer />
        </div>
    );
}

export default App;
