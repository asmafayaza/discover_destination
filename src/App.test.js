import React from 'react';
import ReactDOM from 'react-dom';
import App from './components/App';
import SearchableCityList from './components/SearchableCityList';
import SearchForm from './components/SearchForm';
import SelectDate from './components/SelectDate';
import PassengerCount from './components/PassengerCount';
import reducer from './reducers';
import * as actions from './actions';
import { shallow, mount } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import moment from 'moment';

configure({ adapter: new Adapter() });

const initialState = {};
const mockStore = configureStore();
let store;
beforeEach(() => {
  store = mockStore(initialState)
 })

describe('App Render', () => {
  it('App should render correctly', () => {
    const component = shallow(<App />);

    expect(component).toMatchSnapshot();
  });
})

 describe('SearchForm Render', () => {
   it('SearchForm(Connected Component) should render correctly', () => {
     const component = shallow(<SearchForm store={store}/>)

     expect(component).toMatchSnapshot();
   });
 })


 describe('Test DatePicker', () => {
 it('test should set new date on date change', () => {
     const now = moment();
     const changeDepartureDate = jest.fn();
     let departureDate;
     const wrapper = shallow(<SelectDate changeDepartureDate={changeDepartureDate} selectedDepartureDate={departureDate}/>);
     wrapper.find('.inputDate').first().prop('onChange')(now);
     expect(moment(departureDate).format()).toEqual(moment(now).format());
 });
});

  describe('Test City Selection', () => {
	  it('should call onChange prop', () => {
		const searchOriginCity = jest.fn();
		const event = {
		  target: { value: 'the-value' }
		};
		const component = shallow(<SearchableCityList searchCityList={searchOriginCity} />);
		component.find('.cityInput').first().simulate('change', event);
		expect(searchOriginCity).toBeCalledWith('the-value');
	  });
  })

  describe('Check Disable Button', () => {
    it('Button disable if Passenger more than 5', () => {
      const maxPassengerNo = 6;
      const passengerCount = 6;
      const changePassengerCount = jest.fn();
      const component = shallow(<PassengerCount maxPassenger={maxPassengerNo} passengerCount={passengerCount}/>);
      expect(component.find('#increaseButton').prop('disabled')).toEqual(true);
    });
  })

 describe('Test ActionCreator', () => {
   it('should return expected action', () => {
     const payload = {};
     const expectedAction = {
       type: 'SEARCH_FLIGHT',
       payload
     }
     expect(actions.searchFlight(payload)).toEqual(expectedAction);
   });
 })


 describe('Test Reducer', () => {
   it('should return FlightList', () => {
     const flightList = [
        {
           "id":101,
           "price":"4762",
           "code":"AI-206",
           "origin":"CCU",
           "destination":"BOM",
           "deptTime":"2018-09-27T19:00:00.000",
           "arrTime":"2018-09-27T21:30:00.000"
        },
        {
           "id":102,
           "price":"5064",
           "code":"6E-273",
           "origin":"BOM",
           "destination":"CCU",
           "deptTime":"2018-09-28T02:00:00.000",
           "arrTime":"2018-09-28T04:30:00.000"
        },
        {
           "id":103,
           "price":"4236",
           "code":"AI-204",
           "origin":"CCU",
           "destination":"BOM",
           "deptTime":"2018-09-28T05:00:00.000",
           "arrTime":"2018-09-28T07:30:00.000"
        }
      ]

      const searchString = {
        roundTrip: false,
        originCity: "CCU",
        destinationCity: "BOM",
        departureDate: "2018-09-28T05:00:00.000",
        returnDate: "",
        passengerCount: 1,
        price: {
          min: 1500,
          max: 7000,
        }
      }
      const startAction = {
        type: 'GET_LATEST_FLIGHT_RESULT',
        searchString,
        flightList
      };

      expect(reducer({}, startAction)).toEqual({loadStatus:1, upFlightList:[{
         "id":103,
         "price":"4236",
         "code":"AI-204",
         "origin":"CCU",
         "destination":"BOM",
         "deptTime":"2018-09-28T05:00:00.000",
         "arrTime":"2018-09-28T07:30:00.000"
      }], downFlightList:[]});
   });
 });
