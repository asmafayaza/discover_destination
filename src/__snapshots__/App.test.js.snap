// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Render App should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Connect(SearchForm) />,
        <Connect(FlightList) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Connect(SearchForm) />,
          <Connect(FlightList) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`SearchForm Render SearchForm(Connected Component) should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(SearchForm)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "actions": Object {
        "getLatestFlightResult": [Function],
        "searchFlight": [Function],
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "actions": Object {
          "getLatestFlightResult": [Function],
          "searchFlight": [Function],
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
